### PR TITLE
feat(compiler): Implement constant propagation for template literals

### PR DESCRIPTION
**Pull request by nikeee**
Template literals consisting entirely of constant values will be inlined to a string literal, effectively replacing the backticks with a double quote.

This is done primarily to make the resulting instruction a string literal, so it can be processed further in constant propatation. So this is now correctly simplified to `true`:
```js
`` === "" // now true
`a${1}` === "a1" // now true
```

If a template literal contains a non-constant value, the entire literal is left as-is. Transforming a template literal to one that has only partially inlined is something that can be improved upon in the future.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/ConstantPropagation.ts [hunk 1]
```diff
@@ -472,6 +472,52 @@ function evaluateInstruction(
...
+        }
+        if (subExprValue.kind !== "Primitive") {
+          return null;
+        }
+        if (subExprValue.value === null || subExprValue.value === undefined) {
+          return null;
+        }
+
+        const suffix = value.quasis[quasiIndex].cooked;
+        resultString += subExprValue.value.toString() + suffix;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Optimization/ConstantPropagation.ts:508]: can you double-check the spec behavior here? is it calling toString() or something else? would be great to link to the appropriate point in the spec for posterity.

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/ConstantPropagation.ts [hunk 2]
```diff
@@ -472,6 +472,52 @@ function evaluateInstruction(
...
+
+      let resultString = value.quasis[0].cooked;
+      let quasiIndex = 1;
+      for (const subExprValue of subExpressionValues) {
+        if (!subExprValue) {
+          return null;
+        }
+        if (subExprValue.kind !== "Primitive") {
+          return null;
+        }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Optimization/ConstantPropagation.ts:502]: nit: merge these

**Review by josephsavona — CHANGES_REQUESTED** [review]
My apologies, I overlooked this one. Thanks for the PR! This makes sense and seems reasonable to include, see comments just to make sure the behavior is identical to the spec. We should also test more edge cases: where the interpolation is at the beginning or end of the string, where there are multiple interpolations without any intervening text, with odd values like NaN, empty string, etc.

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/ConstantPropagation.ts (hunk 1 above)
Comment (nikeee) [reply compiler/packages/babel-plugin-react-compiler/src/Optimization/ConstantPropagation.ts:508]: This is a good concern. I've looked it up:
https://262.ecma-international.org/6.0/#sec-template-literals-runtime-semantics-evaluation

It uses string coercion (defined as ToString), wich is defined here: https://262.ecma-international.org/6.0/#sec-tostring
We only deal with string/number/boolean in the case above. So I think the only case that would be non-trivial is `ToString(number)`, which is defined [here](https://262.ecma-international.org/6.0/#sec-tostring-applied-to-the-number-type). [This](https://262.ecma-international.org/6.0/#sec-number.prototype.tostring) is the definition of `Number.prototype.toString()`, which - in case of radix = 10 - just returns `ToString(number)`.

If I'm not mistaking, it matches the spec as long as the values are number/string/booleans. It should be a different story when it is null/undefined (which is excluded via the return above).

In the note it says that the coercion basically behaves like when using [String.prototype.concat](https://262.ecma-international.org/6.0/#sec-string.prototype.concat), which is also the way the TS compiler implements this when lowering to ES3/ES5 (they seem to assume that concat operation is not monkey-patched).

Is this fine or should we just handle the cases where it's trivially always a string? More sophisticated stuff can always be added later.

**Comment by nikeee** [comment]
> We should also test more edge cases: where the interpolation is at the beginning or end of the string, where there are multiple interpolations without any intervening text, with odd values like NaN, empty string, etc.

I agree, will do!
